### DISCUSSION TITLE
Detecting whether or not an operator is valid

### DISCUSSION THREAD

**Discussion by ghost** [sqlalchemy/sqlalchemy#6939] [Usage Questions] [answered]
Hello!

I'm trying to work with a sub-language to create an SQLAlchemy expression, but there is one major issue with input validation. If, for example, a user enters what compiles to `timestamp_column IS TRUE`, an exception will be raised.
Is there a way to detect whether or not a certain operator is valid for a certain column? Something like `"==" in Column.allowed_operators`?
This may very well be outside of SQLAlchemy's scope, though. The exception is only raised at the database driver level.

But regardless, I'll be glad for any input. I've been trying to figure this out for a while now, and validating the user input at an earlier stage would be a huge pain, so I'm exploring this first.

Thanks!

**Comment by huntfx** [comment, marked as answer]
As an alternative, could you use a [custom type](https://docs.sqlalchemy.org/en/14/core/custom_types.html#sqlalchemy.types.TypeDecorator) to offload some of the validation?

**Reply by ghost** [reply]
Thanks for the response!
It's definitely an option, but I'd have to define a custom type for pretty much all my columns then. I think I'd do the validation at an earlier stage, if that was the only option.

**Reply by zzzeek** [reply]
theres not much way to do this without a hardcoded set of rules.   in particular, the "IS" operator is valid for *all* datatypes, since any column of any kind will respond to "column IS NULL".   

I think what you are actually looking for is the casting rules between data **values**, such as if one compares `3 = "3"`.   all the databases have different casting features, sqlite being the most liberal postgresql being the most strict.  to me it seems like a major duplication of effort to anticipate these database-specific behaviors and instead perhaps run the SQL on the target database and then provide friendly parsing / formatting / explanation of the exception messages.

**Reply by ghost** [reply]
I see. Thank you!

( I'd mark your comment as the answer, but it seems Github won't let me do that. Sorry. )
